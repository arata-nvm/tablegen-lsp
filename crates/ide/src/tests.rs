use std::{collections::HashMap, path::Path, sync::Arc};

use crate::{
    db::{SourceDatabase, SourceDatabaseStorage},
    eval::EvalDatabaseStorage,
    file_system::{self, FileId, FilePath, FileSet, FileSystem},
};

const DEFAULT_FILE_PATH: &str = "/main.td";

pub fn single_file(fixture: &str) -> (TestDatabase, Fixture) {
    let mut f = Fixture::single_file(fixture);
    let db = TestDatabase::new(&mut f);
    (db, f)
}

pub fn multiple_files(fixture: &str) -> (TestDatabase, Fixture) {
    let mut f = Fixture::multiple_files(fixture);
    let db = TestDatabase::new(&mut f);
    (db, f)
}

#[salsa::database(SourceDatabaseStorage, EvalDatabaseStorage)]
#[derive(Default)]
pub struct TestDatabase {
    storage: salsa::Storage<Self>,
}

impl salsa::Database for TestDatabase {}

impl TestDatabase {
    fn new(f: &mut Fixture) -> Self {
        let mut db = Self::default();
        for (file_id, content) in f.files() {
            db.set_file_content(file_id, Arc::from(content));
        }

        let source_root = file_system::collect_sources(&mut db, f, f.root_file());
        db.set_source_root(Arc::new(source_root));

        db
    }
}

#[derive(Debug, Default)]
pub struct Fixture {
    file_contents: HashMap<FilePath, String>,
    file_ids: Vec<FileId>,

    file_set: FileSet,
    next_file_id: u32,
}

impl Fixture {
    fn single_file(fixture: &str) -> Self {
        let mut this = Self::default();
        this.insert_file(
            FilePath::from(Path::new(DEFAULT_FILE_PATH)),
            fixture.to_string(),
        );
        this
    }

    fn multiple_files(fixture: &str) -> Self {
        let mut this = Self::default();

        let mut lines = fixture
            .lines()
            .skip_while(|line| !line.starts_with("; "))
            .peekable();
        while let Some(line) = lines.next() {
            let path_str = line
                .strip_prefix("; ")
                .expect("expected header line to start with '; '");
            let path = FilePath::from(Path::new(path_str));

            let mut content = String::new();
            while let Some(line) = lines.peek() {
                if line.starts_with("; ") {
                    break;
                }

                let line = lines.next().expect("line must be present");
                content.push_str(line);
                content.push_str("\n");
            }

            this.insert_file(path, content);
        }

        this
    }

    fn files(&self) -> Vec<(FileId, &str)> {
        self.file_contents
            .iter()
            .map(|(path, content)| {
                (
                    self.file_set
                        .file_for_path(path)
                        .expect("file id must exist"),
                    content.as_str(),
                )
            })
            .collect()
    }

    fn root_file(&self) -> FileId {
        self.file_ids
            .first()
            .expect("at least one file must be exist")
            .clone()
    }

    fn insert_file(&mut self, path: FilePath, content: String) {
        let res = self.file_contents.insert(path.clone(), content);
        assert!(res.is_none(), "duplicate file path");

        let file_id = self.alloc_file_id();
        self.file_set.insert(file_id, path);
        self.file_ids.push(file_id);
    }

    fn alloc_file_id(&mut self) -> FileId {
        let file_id = FileId(self.next_file_id);
        self.next_file_id += 1;
        file_id
    }
}

impl FileSystem for Fixture {
    fn assign_or_get_file_id(&mut self, path: FilePath) -> FileId {
        match self.file_set.file_for_path(&path) {
            Some(file_id) => file_id,
            None => {
                let file_id = self.alloc_file_id();
                self.file_set.insert(file_id, path);
                file_id
            }
        }
    }

    fn path_for_file(&self, file_id: &FileId) -> Option<FilePath> {
        self.file_set.path_for_file(file_id)
    }

    fn read_content(&self, file_path: &FilePath) -> Option<String> {
        self.file_contents.get(file_path).cloned()
    }
}
