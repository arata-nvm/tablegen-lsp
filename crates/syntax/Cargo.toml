[package]
name = "syntax"
version.workspace = true
edition.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
drop_bomb.workspace = true
ecow.workspace = true
insta.workspace = true
rowan.workspace = true
unscanny.workspace = true
