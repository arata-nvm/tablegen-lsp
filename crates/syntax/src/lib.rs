use rowan::ast::AstNode;

use crate::error::SyntaxError;
use crate::lexer::Lexer;
use crate::parser::Parser;
use crate::preprocessor::PreProcessor;
use crate::syntax_kind::SyntaxKind;
use crate::token_kind::TokenKind;

pub mod ast;
pub mod error;
pub mod grammar;
pub mod lexer;
pub mod parser;
pub mod preprocessor;
pub mod syntax_kind;
pub mod token_kind;
pub mod token_stream;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Language {}

impl rowan::Language for Language {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        assert!(raw.0 < SyntaxKind::__LAST as u16);
        unsafe { std::mem::transmute(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        kind.into()
    }
}

pub type SyntaxNode = rowan::SyntaxNode<Language>;
pub type SyntaxToken = rowan::SyntaxToken<Language>;
pub type SyntaxElement = rowan::NodeOrToken<SyntaxNode, SyntaxToken>;

#[derive(Debug, Clone)]
pub struct Parse {
    root_node: SyntaxNode,
    errors: Vec<SyntaxError>,
}

impl Parse {
    pub fn root_node(&self) -> SyntaxNode {
        self.root_node.clone()
    }

    pub fn root(&self) -> Option<ast::Root> {
        ast::Root::cast(self.root_node.clone())
    }

    pub fn errors(&self) -> &[SyntaxError] {
        &self.errors
    }
}

const RECOVER_TOKENS: [TokenKind; 5] = [T![include], T![class], T![def], T![let], T![;]];

pub fn parse(text: &str) -> Parse {
    let lexer = Lexer::new(text);
    let preprocessor = PreProcessor::new(lexer);
    let mut parser = Parser::new(preprocessor, &RECOVER_TOKENS);
    grammar::root(&mut parser);
    let (root_node, errors) = parser.finish();
    Parse { root_node, errors }
}
