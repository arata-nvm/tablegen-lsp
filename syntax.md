SourceFile ::= StatementList
StatementList ::= Statement*
Statement ::= Include | Assert | Class | Def | Defm | Defset | Defvar | Foreach | If | Let | MultiClass

Include ::= "include" String

Class ::= "class" Identifier TemplateArgList? RecordBody

Def ::= "def" Value? RecordBody

Let ::= "let" LetList "in" ( "{" Statement* "}" | Statement )
LetList ::= LetItem ( "," LetItem )*
LetItem ::= Identifier ( "<" RangeList ">" )? "=" Value

MultiClass ::= "multiclass" Identifier TemplateArgList? ParentClassList "{" MultiClassStatement+ "}"
MultiClassStatement ::= Def | Defm | Foreach | Let | If

Defm ::= "defm" Value? ParentClassList ";"

Defset ::= "defset" Type Identifier "=" "{" Statement* "}"

Defvar ::= "defvar" Identifier "=" Value ";"

Foreach ::= "foreach" ForeachIterator "in" ( "{" Statement* "}" | Statement )
ForeachIterator ::= Identifier "=" ForeachIteratorInit
ForeachIteratorInit ::= "{" RangeList "}" | RangePiece | Value

If ::= "if" Value "then" ( "{" Statement* "}" | Statement ) ( "else" ( "{" Statement* "}" | Statement ) )?

Assert ::= "assert" Value "," Value ";"

TemplateArgList ::= "<" TemplateArgDecl ( "," TemplateArgDecl )* ">"
TemplateArgDecl ::= Type Identifier ( "=" Value )?

RecordBody ::= ParentClassList Body
ParentClassList ::= ( ":" ClassRef ( "," ClassRef )* )?
ClassRef ::= Identifier ( "<" ArgValueList? ">" )?
ArgValueList ::= PositionalArgValueList ","? NamedArgValueList
PositionalArgValueList ::= ( Value ( "," Value )* ) ?
NamedArgValueList ::= ( NamedArgValue ( "," NamedArgValue )* )?
NamedArgValue ::= Value "=" Value

Body ::= ";" | "{" BodyItem* "}"
BodyItem ::= FieldDef | FieldLet | Defvar | Assert
FieldDef ::= ( Type | CodeType ) Identifier ( "=" Value )? ";"
CodeType ::= "code"
FieldLet ::= "let" Identifier ( "{" RangeList "}" )? "=" Value ";"

Type ::= BitType | IntType | StringType | DagType | BitsType | ListType | ClassId
BitType ::= "bit"
IntType ::= "int"
StringType ::= "string"
DagType ::= "dag"
BitsType ::= "bits" "<" Integer ">"
ListType ::= "list" "<" Type ">"
ClassId ::= Identifier

Value ::= InnerValue ( "#" InnerValue )*
InnerValue ::= SimpleValue ValueSuffix*
ValueSuffix ::= RangeSuffix | SliceSuffix | FieldSuffix
RangeSuffix ::= "{" RangeList "}"
RangeList ::= RangePiece ( "," RangePiece )*
RangePiece ::= Integer | Integer "..." Integer | Integer "-" Integer | Integer Integer
SliceSuffix ::= "[" SliceElements "]"
SliceElements ::= ( SliceElement "," )* SliceElement ","?
SliceElement ::= Value | Value "..." Value | Value "-" Value | Value Integer
FieldSuffix ::= "." Identifier

SimpleValue ::= Integer | String | Code | Boolean | Uninitialized | Bits | List | Dag | Identifier | ClassValue | BangOperator | CondOperator
Integer ::= INT
String ::= STRING
Code ::= CODE
Boolean ::= "true" | "false"
Uninitialized ::= "?"
Bits ::= "{" ValueList "}"
ValueList ::= Value ( "," Value )*
List ::= "[" ValueList "]"
Dag ::= ( DagArg DagArgList? )
DagArgList ::= DagArg ( "," DagArg )*
DagArg ::= Value ( ":" VARNAME ) | VARNAME
VarName ::= VARNAME
Identifier ::= ID
ClassValue ::= ClassID "<" ArgValueList ">"
BangOperator ::= BANGOP "(" ValueList ")"
CondOperator ::= CONDOP "(" CondClause ( "," CondClause )* ")"
CondClause ::= Value ":" Value
