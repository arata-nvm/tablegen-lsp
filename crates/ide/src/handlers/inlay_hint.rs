use syntax::{
    ast::{self, AstNode},
    parser::TextSize,
    syntax_kind::SyntaxKind,
};

use crate::{
    file_system::FileRange,
    index::IndexDatabase,
    symbol_map::{class::Class, record::RecordField, symbol::Symbol, SymbolMap},
};

#[derive(Debug)]
pub struct InlayHint {
    pub position: TextSize,
    pub label: String,
    pub kind: InlayHintKind,
}

#[derive(Debug)]
pub enum InlayHintKind {
    TemplateArg,
    FieldLet,
}

impl InlayHint {
    fn new(position: TextSize, label: impl Into<String>, kind: InlayHintKind) -> Self {
        Self {
            position,
            label: label.into(),
            kind,
        }
    }
}

pub fn exec(db: &dyn IndexDatabase, range: FileRange) -> Option<Vec<InlayHint>> {
    let index = db.index();
    let symbol_map = index.symbol_map();

    let Some(iter) = symbol_map.iter_symbols_in_range(range) else {
        tracing::info!("no classes found in range: {range:?}");
        return None;
    };

    let mut hints = vec![];
    for (symbol_loc, symbol_id) in iter {
        let symbol = symbol_map.symbol(symbol_id);
        match symbol {
            // TODO: 参照箇所のみをイテレートしたい
            Symbol::Class(class) => {
                if let Some(new_hints) = inlay_hint_class(db, symbol_map, class, symbol_loc) {
                    hints.extend(new_hints);
                }
            }
            Symbol::RecordField(record_field) => {
                if let Some(new_hints) = inlay_hint_record_field(db, record_field, symbol_loc) {
                    hints.extend(new_hints);
                }
            }
            _ => {}
        }
    }
    Some(hints)
}

fn inlay_hint_class(
    db: &dyn IndexDatabase,
    symbol_map: &SymbolMap,
    class: &Class,
    symbol_loc: FileRange,
) -> Option<Vec<InlayHint>> {
    let parse = db.parse(symbol_loc.file);
    let root_node = parse.syntax_node();
    let id_node = root_node.covering_element(symbol_loc.range);
    let identifier_node = match id_node.kind() {
        SyntaxKind::Id => id_node.parent()?,
        SyntaxKind::Identifier => id_node.into_node()?,
        _ => return None,
    };

    let class_node = identifier_node.parent()?;
    let arg_list = match class_node.kind() {
        SyntaxKind::ClassRef => {
            let class_ref = ast::ClassRef::cast(class_node)?;
            class_ref.arg_value_list()?
        }
        SyntaxKind::ClassValue => {
            let class_value = ast::ClassValue::cast(class_node)?;
            class_value.arg_value_list()?
        }
        _ => return None,
    };

    let arg_ranges = arg_list
        .arg_values()
        .take_while(|it| matches!(it, ast::ArgValue::PositionalArgValue(_)))
        .map(|value| value.syntax().text_range());

    let template_arg_names = class
        .iter_template_arg()
        .map(|arg_id| symbol_map.template_arg(arg_id))
        .map(|arg| arg.name.clone());

    let mut hints = vec![];
    for (arg_range, name) in arg_ranges.zip(template_arg_names) {
        hints.push(InlayHint::new(
            arg_range.start(),
            format!("{}:", name),
            InlayHintKind::TemplateArg,
        ));
    }

    Some(hints)
}

fn inlay_hint_record_field(
    db: &dyn IndexDatabase,
    record_field: &RecordField,
    symbol_loc: FileRange,
) -> Option<Vec<InlayHint>> {
    let parse = db.parse(symbol_loc.file);
    let root_node = parse.syntax_node();
    let id_node = root_node.covering_element(symbol_loc.range);
    let identifier_node = match id_node.kind() {
        SyntaxKind::Id => id_node.parent()?,
        _ => return None,
    };

    let maybe_field_let_node = identifier_node.parent()?;
    if maybe_field_let_node.kind() != SyntaxKind::FieldLet {
        return None;
    }

    Some(vec![InlayHint::new(
        symbol_loc.range.end(),
        format!(":{}", record_field.typ),
        InlayHintKind::FieldLet,
    )])
}

#[cfg(test)]
mod tests {
    use crate::tests;

    use super::InlayHint;

    fn check(s: &str) -> Vec<InlayHint> {
        let (db, f) = tests::single_file(s);
        super::exec(&db, f.full_range(f.root_file())).expect("inlay hint not found")
    }

    #[test]
    fn template_arg() {
        insta::assert_debug_snapshot!(check(
            r#"
class Foo<int foo>;
class Bar: Foo<1>;"#
        ));
    }

    #[test]
    fn field_let() {
        insta::assert_debug_snapshot!(check(
            r#"
class Foo {
  int foo;
}
class Bar : Foo {
  let foo = 1;
}
"#
        ));
    }
}
