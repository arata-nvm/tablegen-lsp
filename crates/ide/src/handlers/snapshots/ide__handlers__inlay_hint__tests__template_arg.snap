---
source: crates/ide/src/handlers/inlay_hint.rs
expression: "check(r#\"\nclass Foo<int foo>;\nclass Bar: Foo<1>;\"#)"
---
[
    InlayHint {
        position: 35,
        label: "foo:",
        kind: TemplateArg,
    },
]
